xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -492,9 +493,8 @@ public String toString()
      */
     private Map<DocumentReference, List<BaseObject>> xObjects = new TreeMap<DocumentReference, List<BaseObject>>();
 
-    // TODO: use a Map instead of a List to store attachment in XWikiDocument
     private final List<XWikiAttachment> attachmentList =
-        new AbstractNotifyOnUpdateList<XWikiAttachment>(new ArrayList<XWikiAttachment>())
+        new AbstractNotifyOnUpdateList<XWikiAttachment>(new XWikiAttachmentList())
```

Comment:
Since you introduced a custom class it would be cleaner to also include the features of AbstractNotifyOnUpdateList in it and replace all this code.